stdlib/src/locale.rs

```diff
@@ -0,0 +1,154 @@
+            "locale",
+            "Error",
+            Some(vec![vm.ctx.exceptions.exception_type.to_owned()]),
+        )
+    }
+
+    #[pyfunction]
+    fn localeconv(vm: &VirtualMachine) -> PyResult<PyDictRef> {
+        let result = vm.ctx.new_dict();
+
+        unsafe {
+            let lc = libc::localeconv();
+
+            macro_rules! set_string_field {
+                ($field:ident) => {{
+                    result.set_item(stringify!($field), _parse_ptr_to_str(vm, (*lc).$field)?, vm)?
+                }};
+            }
+
+            macro_rules! set_int_field {
+                ($field:ident) => {{
+                    result.set_item(stringify!($field), vm.new_pyobj((*lc).$field), vm)?
+                }};
+            }
+
+            macro_rules! set_group_field {
+                ($field:ident) => {{
+                    result.set_item(
+                        stringify!($field),
+                        copy_grouping((*lc).$field, vm).into(),
+                        vm,
+                    )?
+                }};
+            }
+
+            set_group_field!(mon_grouping);
+            set_group_field!(grouping);
+            set_int_field!(int_frac_digits);
+            set_int_field!(frac_digits);
+            set_int_field!(p_cs_precedes);
+            set_int_field!(p_sep_by_space);
+            set_int_field!(n_cs_precedes);
+            set_int_field!(p_sign_posn);
+            set_int_field!(n_sign_posn);
+            set_string_field!(decimal_point);
+            set_string_field!(thousands_sep);
+            set_string_field!(int_curr_symbol);
+            set_string_field!(currency_symbol);
+            set_string_field!(mon_decimal_point);
+            set_string_field!(mon_thousands_sep);
+            set_int_field!(n_sep_by_space);
+            set_string_field!(positive_sign);
+            set_string_field!(negative_sign);
+        }
+        Ok(result)
+    }
+
+    #[derive(FromArgs)]
+    struct LocaleArgs {
+        #[pyarg(any)]
+        category: i32,
+        #[pyarg(any, optional)]
+        locale: OptionalArg<Option<PyStrRef>>,
+    }
+
+    #[pyfunction]
+    fn setlocale(args: LocaleArgs, vm: &VirtualMachine) -> PyResult {
+        unsafe {
+            let result = match args.locale {
+                OptionalArg::Missing => {
+                    let null_ptr: *const i8 = ptr::null();
+                    libc::setlocale(args.category, null_ptr)
+                }
+                OptionalArg::Present(locale) => match locale {
+                    None => {
+                        let null_ptr: *const i8 = ptr::null();
+                        libc::setlocale(args.category, null_ptr)
+                    }
+                    Some(l) => {
+                        let mut l_str = l.to_string();
+                        if l_str.is_empty() {
+                            l_str.push('\0');
+                        }
+                        let l_ptr = l_str.as_ptr() as *const i8;
```

Comment:
we prefer to use `std::ffi::CString::new(l)` and `CStr::as_ptr` to obtain the pointer.